runtime/common/src/fees.rs

```diff
@@ -0,0 +1,528 @@
+                    },
+                    Asset::CategoricalOutcome(_, _) |
+                    Asset::ScalarOutcome(_, _) |
+                    Asset::CombinatorialOutcome |
+                    Asset::PoolShare(_) => return Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(6u8))),
+                }
+            }
+
+            fn correct_and_deposit_fee(
+                who: &AccountId,
+                _dispatch_info: &DispatchInfoOf<RuntimeCall>,
+                _post_info: &PostDispatchInfoOf<RuntimeCall>,
+                corrected_native_fee: Self::Balance,
+                _tip: Self::Balance,
+                paid: Self::LiquidityInfo,
+            ) -> Result<(), TransactionValidityError> {
+                let asset_id = paid.asset();
+                let final_fee = match asset_id {
+                    Asset::Ztg => {
+                        // Calculate how much refund we should return.
+                        let (fee, refund) = paid.split(corrected_native_fee);
+
+                        // Refund to the account that paid the fees. If this fails, the account might have dropped
+                        // below the existential balance. In that case we don't refund anything.
+                        let _ = <Tokens as Balanced<AccountId>>::resolve(who, refund);
+
+                        fee
+                    },
+                    #[cfg(not(feature = "parachain"))]
+                    Asset::ForeignAsset(_) => return Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(7u8))),
+                    #[cfg(feature = "parachain")]
+                    Asset::ForeignAsset(_) => {
+                        let (fee_factor, base) = get_fee_factor_and_base(asset_id)?;
+                        let converted_fee = calculate_fee(corrected_native_fee, fee_factor, base)?;
+
+                        // Calculate how much refund we should return.
+                        let (fee, refund) = paid.split(converted_fee);
+
+                        // Refund to the account that paid the fees. If this fails, the account might have dropped
+                        // below the existential balance. In that case we don't refund anything.
+                        let _ = <Tokens as Balanced<AccountId>>::resolve(who, refund);
+
+                        fee
+                    },
+                    Asset::CategoricalOutcome(_, _) |
+                    Asset::ScalarOutcome(_, _) |
+                    Asset::CombinatorialOutcome |
+                    Asset::PoolShare(_) => return Err(TransactionValidityError::Invalid(InvalidTransaction::Custom(8u8))),
+                };
+
+                // Handle the final fee and tip, e.g. by transferring to the treasury.
+                // Note: The `corrected_native_fee` already includes the `tip`.
+                DealWithForeignFees::on_unbalanced(final_fee);
+
+                Ok(())
+            }
+        }
```

Comment:
You can just use [this FungiblesAdapter](https://github.com/paritytech/substrate/blob/dfd82860feb5da6f74ca2aa04414d9cebe310204/frame/transaction-payment/asset-tx-payment/src/payment.rs#L90-L174).

Reply:
https://github.com/zeitgeistpm/zeitgeist/commit/e2e4d21ef6374adf4ce4e3acc9794cfeed8759c6